repair/repair.cc

```diff
@@ -1385,13 +1396,23 @@ future<> repair::user_requested_repair_task_impl::run() {
 
         auto ranges_parallelism = _ranges_parallelism;
         bool small_table_optimization = _small_table_optimization;
-        for (auto shard : boost::irange(unsigned(0), smp::count)) {
+        bool mixed_shard_optimization = _mixed_shard_optimization;
+        unsigned start_shard = 0;
+        unsigned end_shard = smp::count;
+        if (mixed_shard_optimization) {
+            start_shard = utils::uuid_xor_to_uint32(uuid.uuid()) % smp::count;
```

Comment:
How do we ensure we end up iterating over all shards eventually?

Reply:
If there is only one range to repair, only one shard is going to work on that. The uuid is random, so we are replying on the randomness to distributes to all shard overall.

Reply:
Randomness is fine to pick one shard, but we need t have a deterministic algorithm to drive this to completion, covering all shards eventually.

Reply:
Randomness is going to make sure every shard has work to do eventually over time. We do not know in advance what are the next repair requests from user and what they will repair in those requests. We could not act as if we know in advance and distribute them in a round-robin manner to achieve full deterministic.

Reply:
I don't understand. Taking a step back, as a user, how do I use this new flag? Is it enough to just do:
```
curl -X POST http://localhost:1000/storage_service/repair_async/ks?mixed_shard_optimization=true
```
Or will the user have to have a loop of this to drive the repair to completion (repair all shards)?

Reply:
Setting the mixed_shard_optimization=true is enough to turn the flag on and use the feature. But what the user usually does and Scylla manager does is that they also use the ranges parameter to repair only a few ranges per repair request. There will be multiple such requests in practice. Repair will pick a random shard to work on for each such request.